### ISSUE TITLE
[LOW] [Splits] [$500] Web - IOU is greyed out, error message appears in chat and console

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#32076]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.4.0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause - Internal Team
**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/31102
## Action Performed:
1. Navigate to https://staging.new.expensify.com/
2. Log in with a new account
3. Start a chat using the FAB
4. Input an account who isn't registered yet
5. Send a message
6. Upload a picture using the composers "+" button
7. Request money (manual) using the FAB
8. Input an email (can be the one from step 4 or a new one)
9. Finish the IOU

## Expected Result:
I should be able to make the IOU without error messages or console errors

## Actual Result:
I get an error message and the IOU is greyed out. Two errors appear in the console as soon as the error message appears in the chat
Unexpected error creating this chat, please try again later. There is a previously existing chat between these users

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] Windows: Chrome 
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>  

https://github.com/Expensify/App/assets/78819774/a0c7301f-7f2d-463b-b464-05ae860de819

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ae5fe97cb4eccf95</li>
        <li>Upwork Job ID: 1729285304604618752</li>
        <li>Last Price Increase: 2023-11-27</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @iwiznia</details>

**Comment by narefyev91** [comment]
Proposal from @tienifr looks good to me https://github.com/Expensify/App/issues/32076#issuecomment-1828831020
1. We need to add internal BE engineer to work on changes
2. We should assign @tienifr to make draft PR for FE side
3. Combine with BE change and test draft PR to be sure that issue will be resolved
🎀 👀 🎀 C+ reviewed

**Comment by joekaufmanexpensify** [comment]
I'm thinking we should hold off on proceeding with this for right now. We are having some larger internal discussions about this `hidden` feature [here](https://expensify.slack.com/archives/C066HJM2CAZ/p1701194343955309?thread_ts=1701193210.294969&cid=C066HJM2CAZ) that could impact whether we need to actually fix this in this issue.

**Comment by iwiznia** [comment]
ok, putting this on HOLD then. @joekaufmanexpensify you can let us know when we should unhold

**Comment by joekaufmanexpensify** [comment]
Hm, interesting. I'm not sure if creating a chat with one user while online on one device, and then simultaneously attempting a money request with the same user on another device while offline is something a legitimate user would do. 

If the end result of the above discussion is that the hidden case will be fixed by other changes we are making, I'd probably rather wait and see if any actual customers ever experience this other case before doing anything.

**Comment by joekaufmanexpensify** [comment]
Discussion is still ongoing

**Comment by joekaufmanexpensify** [comment]
Same. Looks like we were timing some queries to inform the direction we go in.

**Comment by joekaufmanexpensify** [comment]
Bumped [discussion](https://expensify.slack.com/archives/C066HJM2CAZ/p1702912063800649?thread_ts=1701193210.294969&cid=C066HJM2CAZ) to determine whether anything is needed here.

**Comment by joekaufmanexpensify** [comment]
@iwiznia I [discussed](https://expensify.slack.com/archives/C066HJM2CAZ/p1702912063800649?thread_ts=1701193210.294969&cid=C066HJM2CAZ) this with Puneet more today, and the result of that discussion was that we should proceed with fixing the bug in this issue. 

We are already handling part of this bug [here](https://expensify.slack.com/archives/C066HJM2CAZ/p1701208854452669?thread_ts=1701193218.819059&cid=C066HJM2CAZ) so it works if the actions are all in one session. Meaning, if someone starts a chat with someone they don't know, and then separately creates an IOU with the same person in the same session, we will correctly consolidate the message and IOU into one chat. 

However, we still won't handle [the piece](https://expensify.slack.com/archives/C066HJM2CAZ/p1702916744245259?thread_ts=1701193210.294969&cid=C066HJM2CAZ) related to multiple sessions correctly. Meaning if someone did this, and then signed out and back in, we'd try and put the IOU onto a new chat and it'd error out (as shown above in OP). Puneet mentioned we already handle this case for chats, and we can just re-use that same backend logic for IOUs here.

**Comment by joekaufmanexpensify** [comment]
LMK if you agree! Taking this off of hold for now

**Comment by joekaufmanexpensify** [comment]
Yep, makes sense. My understanding is @iwiznia will handle those in this issue

**Comment by iwiznia** [comment]
> Back-end should return preexistingReportID (similar to OpenReport) if the report creation failure is due to existing report between the 2 users. It should also handle so that the money request is created properly on that preexisting report

In what command?

**Comment by iwiznia** [comment]
ok assigning @tienifr but will not be able to work on the backend of this for a few more days...

**Comment by iwiznia** [comment]
hmmmm looking at the code, I think the RequestMoney does not fail if the report already existed... @tienifr can you provide with the exact response of the failure you are seeing please?

**Comment by joekaufmanexpensify** [comment]
Pending further input from @tienifr

**Comment by joekaufmanexpensify** [comment]
TY!

**Comment by iwiznia** [comment]
In general, please always provide data in text format instead of image, copying that UUID by hand is not fun 😬 

For my own reference, logs are [here](https://www.expensify.com/_devportal/tools/logSearch/#query=request_id%3A%2283f183cf090504c2-HKG%22%20AND%20timestamp%3A%5B2023-12-31T00%3A00%20TO%202024-01-04T23%3A59%5D&index=_all)

**Comment by joekaufmanexpensify** [comment]
I think next steps are on @iwiznia as of now.

**Comment by iwiznia** [comment]
ok looking into this. It seems that `Transaction::createMoneyRequest` throws [here](https://github.com/Expensify/Auth/blob/8d7b81871a91d8abba9e4cbe1eecc394d7c79557/auth/lib/Transaction.cpp#L1738-L1740) when the passed report is different from the one it got from `createChatReport`.
I _think_ I can remove that check and instead add a new piece of code in CreateIOUTransaction to set the `preexistingReportID` parameter to the response in case the ID passed is different than the ID used.
The main questions that arise are:
- Is that ok? Would something break if we remove that check?
- I see createMoneyRequest is used in several other commands (CompleteSplitBill, CreateDistanceRequests, CreateIOUSplit, PayWithWallet, ScrapeCard and SendMoney), I assume this same problem will exist in all the flows that use those and thus want this logic to apply to all of them?

cc @luacmartins since I see you in the blame of the throw.

**Comment by luacmartins** [comment]
Yea, it should be ok to remove that check as long as we update the clients with the correct (pre-existing) reportID. The tricky part is that we also need to clear or update the other optimistic data created, i.e. IOU action and reportPreview to point to the correct reportID too.

**Comment by iwiznia** [comment]
Got it, so how would we go about this? I think the steps would be:
- Update App to handle the new preexistingReportID in all the commands listed above and get it deployed
- Deploy the change to the backend

Right? 
Kind of shitty, since the App will need to be changed without being able to see test with the backend data... but I don't think there's an alternative?

@tienifr that means we would need to update the API commands: CompleteSplitBill, CreateDistanceRequest , SplitBill, SplitBillAndOpenReport, SendMoneyWithWallet., PayMoneyRequestWithWallet, PayWithWallet, SendMoneyElsewhere, SendMoneyViaPaypal to support this.

**Comment by luacmartins** [comment]
I think we can do all of this server side, no? We just need to push an Onyx update to the client to remove the incorrect optimistic data and also send the preexisting report data and the newly created request.

**Comment by iwiznia** [comment]
Oh, can we? I am not really sure how OpenReport is doing that... is it [this code](https://github.com/Expensify/Web-Expensify/blob/bf31f5b1d264c7ce6b9a216f3b5101bdcf4b326f/lib/ReportAPI.php#L5569-L5582) and there's really nothing needed to be changed in App?

**Comment by luacmartins** [comment]
Yea, that's the code that handles that. That being said, I just looked at the OP and I take back what I said before. This seems to be an issue with `RequestMoney` not getting the existing chat report between the users on the client and then generating an optimistic one.

**Comment by iwiznia** [comment]
Sorry, are you agreeing that we do need to update all the commands I listed above or something else?

**Comment by luacmartins** [comment]
It doesn't seem like we do. I didn't read the whole conversation here, but from the video in the OP, I _think_ the issue is that App is not getting the existing chatReport and is creating a brand new one, which is incorrect

**Comment by iwiznia** [comment]
I don't think that's the error... the backend should return the existing report if one exists because it's possible the frontend took the action while offline, where it is expected to not have all reports loaded.

**Comment by iwiznia** [comment]
@tienifr thoughts?

**Comment by joekaufmanexpensify** [comment]
Still pending input from @tienifr

**Comment by luacmartins** [comment]
Yea, I agree with that, but I thought the backend already handled that [here](https://github.com/Expensify/Web-Expensify/blob/bf31f5b1d264c7ce6b9a216f3b5101bdcf4b326f/lib/ReportAPI.php#L5569-L5582). Maybe there's a different flow 🤷 But yes, in that case the backend should handle updating the data with the correct reportID.

**Comment by joekaufmanexpensify** [comment]
It seems like we're still determining best next steps for this one. Is that right?

**Comment by joekaufmanexpensify** [comment]
Awaiting confirmation on whether more discussion is needed to proceed with fixing this one

**Comment by joekaufmanexpensify** [comment]
@iwiznia is there any more discussion needed before we proceed with fixing this one?

**Comment by joekaufmanexpensify** [comment]
Bump @iwiznia

**Comment by joekaufmanexpensify** [comment]
Pending further input

**Comment by joekaufmanexpensify** [comment]
@tienifr @iwiznia @luacmartins is any further discussion needed here before we can begin working on this?

**Comment by luacmartins** [comment]
I think we agree that this is a backend issue and should be internal. Not sure if there's anything else to clarify at this point.

**Comment by joekaufmanexpensify** [comment]
Bumped Ioni 1:1

**Comment by joekaufmanexpensify** [comment]
Still pending further input

**Comment by joekaufmanexpensify** [comment]
@iwiznia updating you to be the owner of this one, as my understanding is this needs to be backend, and we should be good to get started. LMK if that is not correct though!

**Comment by joekaufmanexpensify** [comment]
Bump @tienifr @luacmartins when you have a sec

**Comment by iwiznia** [comment]
@luacmartins do you agree with [this](https://github.com/Expensify/App/issues/32076#issuecomment-1936081967)?

**Comment by luacmartins** [comment]
Yes, that sounds correct to me.

**Comment by iwiznia** [comment]
No update

**Comment by iwiznia** [comment]
No update

**Comment by iwiznia** [comment]
No update

**Comment by iwiznia** [comment]
No update

**Comment by iwiznia** [comment]
No update

**Comment by iwiznia** [comment]
No update

**Comment by iwiznia** [comment]
No update

**Comment by joekaufmanexpensify** [comment]
Hm, I don't think it is explicitly on hold. It's a question of internal priorities, given the next step here is for @iwiznia to put up the backend fix discussed [here](https://github.com/Expensify/App/issues/32076#issuecomment-1936081967). Defer to him on when he thinks he'll be able to prioritize it. Happy to move it to weekly if that will be soon.

**Comment by iwiznia** [comment]
The changes needed are quite large and involve lots of api commands... given the low severity of the issues and the fact that reproduction steps are quite complex, I think we might be better off closing it.
